xarray/namedarray/utils.py

```diff
@@ -80,6 +78,54 @@ def to_0d_object_array(
     return result
```

Comment:
This has been changed in a different PR, but while looking at the changes I noticed that without involving a type checker `is_duck_dask_array` returns the same result as `is_dask_collection`. Could we also do a runtime check that `x` is a duckarray? (`isinstance(x, _arrayfunction_or_api)` if I understand correctly)

Reply:
I don't think it's necessary if we agree that these functions are private. 
All testing in xarray and NamedArray will use a type checker and will fail if anything else than a duckarray is input.

Reply:
You have a point: the only place we use it for at the moment is in the `__dask_*__` protocols where we can guarantee that whatever we pass is a duck array.

However, the fact remains that without a type checker the function name is misleading: it will not check that whatever is passed is a duck array, and any future contribution will have to be aware of that quirk. So instead, I'd rather use `is_dask_collection` directly in the `__dask_*__` protocols, because that makes the fact that we don't check for duck arrays again explicit.

Reply:
There's a lot of quirks hidden in the pytests as well, future contributors will have to deal with those (and mypy) in the CI if they want something merged. 
Just like we are weary of removing tests without motivation we should be weary of using `# type: ignore` or `Any` without motivation. It's another tool I think we should use more internally to speed things up.

`is_duck_dask_array` and `is_dask_collection` are typeguarded differently, so they have a little different flavour. Maybe using is_dask_collection only is good enough.